File: set1/ch4/break_file.c
#include <stdio.h>
#include <string.h>
#include <float.h>
#include <math.h>
#include <stdlib.h>
//#include "single_byte_xor.h"

char alphabet[53] = {'a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z','A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z',' '};

// probabilities source from http://pi.math.cornell.edu/~mec/2003-2004/cryptography/subs/frequencies.html
float freq_table[27] = {8.12,1.49,2.71,4.32,12.02,2.30,2.03,5.92,7.31,0.10,0.69,3.98,2.61,6.95,7.68,1.82,0.11,6.02,6.28,9.10,2.88,1.11,2.09,0.17,2.11,0.07,18.29};

int hti(unsigned char hex) {
	if ((hex > 96) && (hex < 103))
		return hex - 87;
	else
		return hex - 48;
}

char dec(char a, char b) {
	a = hti(a); b = hti(b);
	char c = a << 4;
	return (c |= b);
}

double comp_exp(int c, int len, double n_obs) {
	if ((c > 96) && (c < 123)) {
		c -= 97;
		return freq_table[c] * len;		
	} else if ((c > 64) && (c < 91)) {
		c -= 65;
		return freq_table[c] * len;
	} else if (c == 32) {
		c -= 5;
		return freq_table[c];
	} else
		return n_obs;
}

double comp_obs(char c, char string[]) {
	double sz = (double)strlen(string);
	double acc = 0.0;
	for (int i = 0; i < sz; i++) {
		if (string[i] == c)
			acc++;
	}
	return acc / sz;
}

double sc(char newcode[]) {
	double acc = 0.0;
	int sz = (int)strlen(newcode);
	for (int i = 0; i < sz; i++) {
		double n_obs = comp_obs(newcode[i], newcode);
		double n_exp = comp_exp(newcode[i], sz, n_obs);
		if (n_exp == 0.0)
			continue;
		double diff = (n_obs - n_exp);
		acc += ((diff*diff) / n_exp);
	}
	return acc;
}

char ciph(char code[], char plaintext[], char x) {
	int sz = (int)strlen(alphabet);
	int clen = (int)strlen(code);
	float score = 0.0;
	float sc_code = 0.0;
	char key = 0;
	for (int i = 0; i < sz; i++) {
		int k = 0;
		char* newcode = malloc(clen);
		for (int j = 0; j < clen; j+=2) {
			newcode[k] = dec(code[j], code[j+1]) ^ alphabet[i];
			k++;
		}
		sc_code = sc(newcode);
		//printf("%f\n", sc_code);
		if (sc_code > score) {
			//printf("Char %c for %d: %s\n", alphabet[i], x, newcode);
			score = sc_code;
			key = alphabet[i];
			strncpy(plaintext, newcode, clen);
		}
		free(newcode);
	}
	return key;
}

int main(void) {
	FILE* fp = fopen("file.txt", "r");
	char buffer[62];
	char plaintext[62] = {0};
	char splain[62] = {0};
	double min = -100.0;
	double score = 0.0;
	int i = 0;
	int n = 0;
	char key = 0;
	char k = 0;
	while (fgets(buffer, 62, fp) != NULL) {
		key = ciph(buffer, plaintext, i);
		score = sc(plaintext);
		printf("%s\n", plaintext);
		//printf("%f\n", score);
		if (score > min) {
			min = score;
			strncpy(splain, plaintext, 62);
			k = key;
			n = i;
		}
		i++;
	}
	printf("Plaintext: %s\n", splain);
	printf("line: %d\n", n);
	printf("key: %c\n", k);
	return 0;
}

File: set1/ch4/break_file.py
from collections import defaultdict

alphabet = ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z','A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z',' ']

freq_table = [8.12,1.49,2.71,4.32,12.02,2.30,2.03,5.92,7.31,0.10,0.69,3.98,2.61,6.95,7.68,1.82,0.11,6.02,6.28,9.10,2.88,1.11,2.09,0.17,2.11,0.07,18.29]

def hti(h):
	if ((h > 96) and (h < 103)):
		return h - 87
	return h - 48

def decode(a, b):
	a = hti(ord(a))
	b = hti(ord(b))
	a = a << 4
	return a | b

def compute_exp(c, n_obs, l):
	#print(c)
	if ((c > 96) and (c < 123)):
		c -= 97
		return freq_table[c] * l
	elif ((c > 64) and (c < 91)):
		c -= 65
		return freq_table[c] * l
	elif (c == 32):
		c -= 6
		return freq_table[c]
	else:
		return n_obs

def compute_obs(c, string):
	sz = len(string)
	acc = 0.0
	for i in string:
		if (i == c):
			acc+=1
	return acc / sz

def score_code(newcode):
	sz = len(newcode)
	acc = 0.0
	for i in newcode:
		#print(f"newcode[i]: {newcode[i]}")
		n_obs = compute_obs(i, newcode)
		#print(f"n_obs: {n_obs}")
		n_exp = compute_exp(ord(i), n_obs, len(newcode))
		#print(f"n_exp: {n_exp}")
		#print(f"n_obs: {n_obs}, n_exp: {n_exp}")
		if (n_exp == 0):
			continue
		n = (n_obs - n_exp)
		#print(f"n: {n}")
		
		n2 = n*n
		#print(f"n2: {n2}")
		
		#print(f"n: {n}")
		res = (n2 / n_exp)
		#print(f"res: {res}")
		
		#print(res)
		acc += res
	#print(f"acc: {acc}")
	return acc

def break_cipher(code):
	sz = len(alphabet)
	clen = len(code)-1
	score = 0.0
	sc = 0.0
	plaintext = ''
	b = defaultdict(lambda: 0.0)
	for i in range(0, sz):
		newcode = []
		for j in range(0, clen//2, 2):
			#print(f"{code[j]}{code[j+1]}")
			d = decode(code[j], code[j+1])
			#print(f"d: {d}")
			newcode += chr(d ^ ord(alphabet[i]))
		#print(f"newcode: {newcode}")
		sc = score_code(newcode)
		if sc > score:
			b[''.join(newcode)] = score
			score = sc
			plaintext = newcode
	dd = sorted(b.items(), key=lambda x:x[1])
	for e in dd:
		print(e)
	return plaintext

def main():
	f = open('file.txt','r')
	d = defaultdict(lambda:0.0)
	for line in f:
		decoded = break_cipher(line)
		score = score_code(decoded)
		decoded = ''.join(decoded)
		d[decoded] = score
		#print(f"line: {decoded} score: {score}")
	#dd = sorted(d.items(), key=lambda x:x[1])
	#for e in dd:
	#	print(e)
	#f.close()

main()
